File: Cpp/137.cpp
// @Date       : 2022/1/31 
// @Filename   : 137.cpp
// @Tag        :
// @Autor      : LI YAO
// @Difficulty :

#include <iostream>
#include <string>
#include <vector>

using namespace std;

// ----------------
class Solution {
public:
	int singleNumber(vector<int>& nums) {
		int ans = 0;
		for (int i = 0; i< 32; ++i){
			int total = 0;
			for (int num: nums){
				total += (num >> i) & 1;
			}
			
			if (total % 3){
				ans |= (1 << i);  // 或等于
			}
			
		}
		return ans;
	}
};


//------------------


int main(){
	Solution a;
	vector<int> b = {1,1,1,2,4,4,4,5,5,5,3,3,3};
	cout<<a.singleNumber(b);
	
}

File: Python/137.py
#!/usr/bin/env python3
# @Date       : 2022/1/31 
# @Filename   : 137.py
# @Tag        :
# @Autor      : LI YAO
# @Difficulty :

from heapq import *
from typing import List,  Optional
from collections import defaultdict, deque, Counter
from itertools import product,combinations,permutations,accumulate
from random import choice, randrange,randint

class ListNode: 
	def __init__(self, val=0, next=None):
		self.val = val
		self.next = next

# -------------------------
# 哈希表
#class Solution:
#	def singleNumber(self, nums: List[int]) -> int:		
#		c = Counter(nums)
#		for i, j in c.items():
#			if j == 1:
#				return i
		
# -------------------------
# 位运算，计算出每一位的和，然后对3取余
class Solution:
	def singleNumber(self, nums: List[int]) -> int:		
		ans = 0
		for i in range(32):
			total = sum((num>>i) & 1 for num in nums)
			if total % 3:
				if i == 31:
					ans -= (1 << i)
				else:
					ans |= (1<<i)
		return ans
		
		
# --------------------------
a = Solution()